writeengine/wrapper/we_colop.cpp

```diff
@@ -1791,6 +1800,59 @@ int ColumnOp::writeRows(Column& curCol, uint64_t totalRow, const RIDList& ridLis
   return rc;
 }
 
+/***********************************************************
+ * DESCRIPTION:
+ *    MCOL-5021 Read-only version of writeRows() function.
+ * PARAMETERS:
+ *    curCol - column information
+ *    totalRow - the total number of rows that need to be read
+ *    ridList - the vector of row id
+ *    oldValArray - the array of old value
+ * RETURN:
+ *    NO_ERROR if success, other number otherwise
+ ***********************************************************/
+int ColumnOp::writeRowsReadOnly(Column& curCol, uint64_t totalRow, const RIDList& ridList,
+                                void* oldValArray)
+{
+  uint64_t i = 0, curRowId;
+  int dataFbo, dataBio, curDataFbo = -1;
+  unsigned char dataBuf[BYTE_PER_BLOCK];
+  bool bExit = false;
+  int rc = NO_ERROR;
+
+  while (!bExit)
+  {
+    curRowId = ridList[i];
+
+    calculateRowId(curRowId, BYTE_PER_BLOCK / curCol.colWidth, curCol.colWidth, dataFbo, dataBio);
+
+    // load another data block if necessary
+    if (curDataFbo != dataFbo)
+    {
+      curDataFbo = dataFbo;
+      //@Bug 4849. need to check error code to prevent disk error
+      rc = readBlock(curCol.dataFile.pFile, dataBuf, curDataFbo);
+
+      if (rc != NO_ERROR)
+        return rc;
+    }
+
+    // Read the old value of the record
+    if (oldValArray)
+    {
+      uint8_t* p = static_cast<uint8_t*>(oldValArray);
+      memcpy(p + i * curCol.colWidth, dataBuf + dataBio, curCol.colWidth);
+    }
+
+    i++;
+
+    if (i >= totalRow)
```

Comment:
Do not get how `i` could be greater than `totalRow`, you increment `i` at the line 1847, the `totalRow` has const semantics in this function.

Reply:
Good catch, the boolean var to check the loop condition was not required, so I have removed it.